internal/controller/telemetry/route_controller.go

```diff
@@ -0,0 +1,399 @@
+			return nil, err
+		}
+
+		namespaces = append(namespaces, namespacesForSelector.Items...)
+	}
+
+	namespaces = normalizeNamespaceSlice(namespaces)
+
+	return namespaces, nil
+}
+
+// disownSubscriptions fails internally by logging errors individually
+// this is by design so that we don't fail the whole reconciliation when a single subscription update fails
+func (r *RouteReconciler) disownSubscriptions(ctx context.Context, subscriptionsToDisown []v1alpha1.Subscription) {
+	logger := log.FromContext(ctx)
+	for _, subscription := range subscriptionsToDisown {
+		subscription.Status.Tenant = ""
+		err := r.Client.Status().Update(ctx, &subscription)
+		if err != nil {
+			logger.Error(err, fmt.Sprintf("failed to detach subscription %s/%s from collector", subscription.Namespace, subscription.Name))
+		} else {
+			logger.Info("disowning subscription", "subscription", fmt.Sprintf("%s/%s", subscription.Namespace, subscription.Name))
+		}
+	}
+}
+
+// updateSubscriptionsForTenant fails internally and logs failures individually
+// this is by design in order to avoid blocking the whole reconciliation in case we cannot update a single subscription
+func (r *RouteReconciler) updateSubscriptionsForTenant(ctx context.Context, tenantName string, subscriptions []v1alpha1.Subscription) (updatedSubscriptions []v1alpha1.Subscription) {
+	logger := log.FromContext(ctx, "tenant", tenantName)
+	for _, subscription := range subscriptions {
+		subscription.Status.Tenant = tenantName
+
+		logger.Info("updating subscription status for tenant ownership")
+		err := r.Status().Update(ctx, &subscription)
+		if err != nil {
+			logger.Error(err, fmt.Sprintf("failed to set subscription (%s/%s) -> tenant (%s) reference", subscription.Namespace, subscription.Name, tenantName))
+		} else {
+			updatedSubscriptions = append(updatedSubscriptions, subscription)
+		}
+	}
+	return
+}
+
+func (r *RouteReconciler) getSubscriptionsReferencingTenantButNotSelected(ctx context.Context, tenant *v1alpha1.Tenant, selectedSubscriptions []v1alpha1.Subscription) []v1alpha1.Subscription {
+	logger := log.FromContext(ctx)
+	var subscriptionsReferencing v1alpha1.SubscriptionList
+	listOpts := &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(tenantReferenceField, tenant.Name),
+	}
+
+	if err := r.Client.List(ctx, &subscriptionsReferencing, listOpts); client.IgnoreNotFound(err) != nil {
+		logger.Error(err, "failed to list subscriptions that need to be detached from tenant")
+		return nil
+	}
+
+	var subscriptionsToDisown []v1alpha1.Subscription
+
+	for _, subscriptionReferencing := range subscriptionsReferencing.Items {
+		selected := false
+
+		for _, selectedSubscription := range selectedSubscriptions {
+			if subscriptionReferencing.Name == selectedSubscription.Name {
+				selected = true
+				break
+			}
+		}
+
+		if !selected {
+			subscriptionsToDisown = append(subscriptionsToDisown, subscriptionReferencing)
+		}
```

Comment:
[nit]
I think this would be shorter with `slices.IndexFunc`, what do you think?